### PR TITLE
Fixes #3101: Break JSX attributes onto their own lines if any newlines are present.

### LINKED ISSUE #3101: Option to retain multi-line props in JSX/HTML
Right now I'm finding that Prettier collapses JSX attributes onto a single line, up until the column width. This (in my opinion) feels much harder to read as it requires the eyes to constantly scan left to right rather than "glance" at a single block. 

Would you be open to an option that disables this forced single-line output? 

**Prettier 1.7.4**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22const%20f%20%3D%20()%20%3D%3E%20(%5Cn%20%20%3CFoo%5Cn%20%20%20%20hello%3D%7Bbaz%7D%5Cn%20%20%20%20how%3D%7Bare%7D%5Cn%20%20%20%20you%3D%7Ba%7D%5Cn%20%20%20%20hi%3D%7Bc%7D%5Cn%20%20%2F%3E%5Cn)%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)

**Input:**
```jsx
const f = () => (
  <Foo
    hello={baz}
    how={are}
    you={a}
    hi={c}
  />
)
```

**Output:**
```jsx
const f = () => <Foo hello={baz} how={are} you={a} hi={c} />;
```

**Expected behavior:**
Retain line breaks. 

### Update: 

I opened this a long while back but after two years of using Prettier I absolutely, without question, and without hestitation, support the settings currently in Prettier and further -- believe that all of these personal preferences simply vanish from sight after using Prettier for more than a few days, as if one never even had them. They vanish into the delight of not having to worry about preferences anymore, and all of the energy / time savings that come from that. I am a 👎 on adding additional configuration in this regard.

### PR DESCRIPTION
**Pull request by seansfkelley**
Per https://github.com/prettier/prettier/issues/3101#issuecomment-339132007, loosely based on #495.

### CODE REVIEW

File: tests/jsx/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -571,7 +571,9 @@ exports[`expression.js - flow-verify 1`] = `
   value={option}
 />;
 
-<ParentComponent prop={<Child>test</Child>} />;
+<ParentComponent
+  prop={<Child>test</Child>}
+/>;
```
Comment (seansfkelley) [inline tests/jsx/__snapshots__/jsfmt.spec.js.snap]: The original text here was

```jsx
<ParentComponent prop={
  <Child>
    test
  </Child>
}/>;
```

This is another weird case like https://github.com/prettier/prettier/pull/4862/files#r203585057.

File: tests/flow/new_react/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -530,7 +535,9 @@ var C = React.createClass({
   }
 });
 
-<C statistics={[{}, { label: "", value: undefined }]} />; // error (label is required, value not required)
+<C
+  statistics={[{}, { label: "", value: undefined }]}
+/>; // error (label is required, value not required)
```
[the remark this answers is not in the document]
Comment (j-f1) [reply tests/flow/new_react/__snapshots__/jsfmt.spec.js.snap]: What about only breaking if the linebreaks are *between* the props?

**Comment by duailibe** [comment]
> JSX attribute lists are not that different from large object literals

Except they are. From https://github.com/prettier/prettier/issues/2668#issuecomment-324990700

> The reason why we keep the objects expanded is because objects in JavaScript can mean different things: classes, hashmaps, configuration, named arguments, returning a tuple... We want to be able to print those use cases in different ways but unfortunately we haven't found a good way to automatically figure out which is which based on the AST.
>
> ES6, Flow and TypeScript have been awesome for prettier because it added a bunch of syntax to differentiate those different use cases.
>
> The decision to keep the newlines around in objects has not been taken lightly because it chips away the value of not having to care about how you format your code. This is an exception and has caused a bunch of confusion already.
>
> JSX doesn't have this wide range of use cases so we don't want to apply the same technique.

**Comment by seansfkelley** [comment]
While it's true that JSX attributes aren't used to define classes and tuples, the other three options presented (hashmaps, configuration and named arguments) are all legitimate use-cases. Plus, JSX is just syntactic sugar on actual object literals. It's just that the top level of the literal is spelled `={}` instead of `: `.

Here's an example from some actual code I wrote before formatting:

```tsx
<ButtonedListGroupItem
  key={key}
  className="clean"
  buttons={<>
    <IconButton 
      title="Save Edits"
      icon="save"
      family="r"
      onClick={this.save}
    />
    <IconButton 
      title="Cancel Editing"
      icon="ban"
      onClick={this.cancel}
    />
  </>}
/>
```

[Prettier with default settings thinks this should be](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAQgVxjacAmAMgJYDOMA4gE4SYAOAkvALYA6UABBwNZwCeAXmC8+AX3ZcwAGwCGpUgDkZzOANbgpcGVHUSOAI2y4opIagB8erqgaQoWHNA5WuHGMRia1IAMoyAbnAcAKL4HqS6nK5cxHbepAFwkdFcAGbKxFKC6lTJKdAAwlKx3EIwABZkAHQJgeJRXAD0lg0cNnYOxs6tXO6equoF2mBwUiFh7lAA5nnRsdDe+tqzroXFYKXAFdVgw6P10c16qM31zSAANCAQdO7QpMigMlQ0AO4ACs8IDygBEMT4S4gfRUGQbOAwHx0MHEabIGBUTBwK6w0hwKgwd6gqbMGTIdJSNFXABWpAAHuhQeDIco4CQoHB8TJCciQNCqGiqMhgTJ9HwpNAgXQqLCYAB1AEVZAADgADFdhRA0WLQXRucK4JzAkCqHAAI6YYi6rEyHF4pAEokgNHMYjwxGs0iwqaaACKmAg8CZLKuMF5EvwUqQACZfaDMs6ChBmLjuVA8EDMGiACq8n6WuCiURAA):

```tsx
<ButtonedListGroupItem
  key={key}
  className="clean"
  buttons={
    <>
      <IconButton
        title="Save Edits"
        icon="save"
        family="r"
        onClick={this.save}
      />
      <IconButton title="Cancel Editing" icon="ban" onClick={this.cancel} />
    </>
  }
/>;
```

Which is clearly worse. Like https://github.com/prettier/prettier/issues/3101#issuecomment-387875462 and https://github.com/prettier/prettier/issues/3101#issuecomment-387898876, I'm trying to call out while authoring that there is a consistency here that can be usefully captured with consistent use of whitespace. Prettier moved the fragment around and such, that's fine, but it throws out this information that I have chosen to provide.

Let's condense this a bit and rewrite it with `React.createElement`:

```ts
React.createElement(ButtonedListGroupItem, {
  key,
  className: "clean",
  buttons: React.createElement(React.Fragment, {}, 
    React.createElement(IconButton, { 
      title: "Save Edits",
      icon: "save",
      family: "r",
    }),
    React.createElement(IconButton, { 
      title: "Cancel Editing",
      icon: "ban",
    }),
  ),
})
```

It's not very different at all. And [Prettier will output something nice for this](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAlOBDMMB0YBOm8AogDZwC2CMAFAEICuMM0cAJgDICWAzjAOL4IDAA4BJeBQA0AAmAAdKDJkBrOAE8pi5WFIYePAHIYqSGfPDkMUC1qUyARkxZQeZ9FlwEicMpWo0Htg4AGL4GADmVLCywAC+strKMkFehBgk5NG0YpBQjMzQseb2yTIwXDDkZhYAyhgAbnAyxGyVPLZJZVx5NSA8jXCdpckAZiZcpOp9+MNlcQCUdmWpeOmZ-rA0udAFLsVdyRVVcH0AwtZgcKQtbRVQEXNlMj3QfQ7WT8qLyzJLiosQFIQBARBVoDxkKAMPghAB3AAKMIQkJQjQgXDYQJADnCYDUMFqIiwXAeyBg+AYcGBpJ4cHwMAR4SiGGQ41IdOBACseAAPOh4gn1KjcKBwNkYDnUkDE-B0-DIHEYBzqUjQbEifCkmAAdUxMAAFsgABwABmBmogdJ14REis1cHlTWxhAAjgwuIQmZEKKykOzOf1KFxyZTpTxSRFyABFBgQeASqXAmDKvVsQ3IABMyfCk0jZwgFF9iqgrGxDDpABVlaiA3A4nEgA):

```ts
React.createElement(ButtonedListGroupItem, {
  key,
  className: "clean",
  buttons: React.createElement(
    React.Fragment,
    {},
    React.createElement(IconButton, {
      title: "Save Edits",
      icon: "save",
      family: "r"
    }),
    React.createElement(IconButton, {
      title: "Cancel Editing",
      icon: "ban"
    })
  )
});
```

But if, for some reason, I am compelled to one-line the props:

```ts
React.createElement(ButtonedListGroupItem, {
  key,
  className: "clean",
  buttons: React.createElement(React.Fragment, {}, 
    React.createElement(IconButton, { title: "Save Edits", icon: "save", family: "r" }),
    React.createElement(IconButton, {  title: "Cancel Editing", icon: "ban" }),
  ),
})
```

[Prettier will rearrange only part of it](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAlOBDMMB0YBOm8AogDZwC2CMAFAEICuMM0cAJgDICWAzjAOL4IDAA4BJeBQA0AAmAAdKDJkBrOAE8pi5WFIYePAHIYqSGfPDkMUC1qUyARkxZQeZ9FlwEicMpWo0Htg4AGL4GADmVLCywAC+strKMkFehBgk5NG0YpBQjMzQsTIwXDDkZhYAyhgAbnAyxGxlPLYyXHmVIDx1cG0AZiZcpOpd+BYycQCUdskpmMHeGb5ZAbnQBS7FJWUV5iAAwtZgcKSNzaVQEW0d0F0O1hPTszIzitMgUiAQIqXQPMhQBh8EIAO4ABWBCABKDqEC4bE+IAc4TAahgVREWC4V2QMHwDDgXxxPDg+Bg4PCUQwyEGpFJXwAVjwAB50VHompUbhQOC0jD0okgLH4Un4ZDIjAOdSkaBIkT4HEwADqCJgAAtkAAOAAMXwVEFJyvCIglCrgYvqSMIAEcGFxCJTIhQaUg6QzupQuHiCUKeDiIuQAIoMCDwfmCr4wKWqtga5AAJij4WGAYOEAoLolUFYSIYpIAKlKYe64HE4kA):

```ts
React.createElement(ButtonedListGroupItem, {
  key,
  className: "clean",
  buttons: React.createElement(
    React.Fragment,
    {},
    React.createElement(IconButton, {
      title: "Save Edits",
      icon: "save",
      family: "r"
    }),
    React.createElement(IconButton, { title: "Cancel Editing", icon: "ban" })
  )
});
```

Which, by virtue of #495, we have already determined is undesirable. I argue that JSX should be treated the same in these cases as object literals. Plus, I'm trying to respect the option philosophy here to do the right thing and produce nice results rather than add extra knobs. I like the idea that it's opinionated, I just think it could be a little more self-consistent (and prettier). :)

**Comment by duailibe** [comment]
Fair enough.. I guess this could also benefit of @j-f1's idea of:

- always break if it doesn't fit
- if props > N, preserve original breaks
- else (i.e. props <= N), preserve current formatting

WDYT? cc @j-f1

**Comment by j-f1** [comment]
Perhaps set N to two in that case @duailibe? We shouldn’t be breaking for a single prop if it would fit on one line.

**Comment by seansfkelley** [comment]
So to be explicit, you're suggesting that we use one-line format even if there are user-entered line breaks in the case where there is a single prop. Correct?

I'm open to this idea, but here's a situation in which it may not be desirable:

```tsx
// I like to break my react-router Route components out onto multiple lines even if they
// have one prop, because they are often next to other Routes that have multiple props.
<Route 
  path="/" 
  exact 
  render={this.renderSitesRedirect}
/>
<Route 
  render={this.renderHeader}
/>
<Route 
  render={this.renderBody}
/>
```

I don't feel too strongly about this particular case and I'm sure there is some heuristic fiddling which will get a "generally nice" result. N = 2 is probably best if we go down this path.

On the topic of deciding what it means for the user to have already included a line break, do you figure the current version ("there exists a newline anywhere in this text span") is sufficient, or we want to be a little more clever ("there exists a prop name that is the first non-whitespace member on its line")?

**Comment by duailibe** [comment]
I was actually thinking of N = 3.. for N = 2 it would just be "preserve line breaks unless it's a single prop" :P

**Comment by seansfkelley** [comment]
I'm down with the current line break detection behavior.

@j-f1 thoughts on values of N? I'd prefer smaller values in deference to the "hinting" provided by the user-author (such as... 0 😄 ), which would mimic object literals more closely.

**Comment by seansfkelley** [comment]
Ping @duailibe, @j-f1! I'm happy with this code as-is and it sounds like we're at or near agreement. Is there something I can do carry this PR through?

**Comment by vjeux** [comment]
I'm sorry to be a kill joy but this is not the fix that we should do in this case. The root cause of the problem here is that if you have multiple elements next to each other that have a similar shape, you want them to be expanded a similar way and not having one which is inline because it happens to fit 80 columns.

If you want to solve this problem, we need to find a way to detect this pattern and address it correctly. So far I don't know of a great way of doing that but that's worth investigating.

The solution proposed here is going to make code formatted in different ways based on the way you type it. So it makes a dent in the "just write everything in one line and save" feeling. And also now that you have a choice, people may ask you to do it a certain way in pull requests.

We've done it for objects because objects are used in a wide variety of contexts in JavaScript. For example React.createClass({}) is really a class so you always want to break. But it's almost impossible to tell from just the syntax what the intention was.

This was also a case where people would straight up refuse to use prettier in that state and the vast majority of the issues opened were due to this. While this problem here is unfortunate, it is way less problematic than the issue with object was and I believe doesn't require such drastic measure.

**Comment by suchipi** [comment]
I agree with @vjeux, my opinion of this has changed since the linked comment I posted.

**Comment by seansfkelley** [comment]
#3101 has been stalled for a while so I'm trying to get the ball rolling on it with a specific proposal.

> If you want to solve this problem, we need to find a way to detect this pattern and address it correctly. 

If this solution is not acceptable, what is? Can you outline an algorithm in plain English that would have the desired result without violating the principles you cite?

> The solution proposed here is going to make code formatted in different ways based on the way you type it. So it makes a dent in the "just write everything in one line and save" feeling. And also now that you have a choice, people may ask you to do it a certain way in pull requests.

To be clear, given the stated goals of the project, I think this is an accurate criticism though (as I pointed out in comments before) I think it constitutes putting the cart before the horse. But above all I want this feature to exist and work so... how should we handle it?

**Comment by vjeux** [comment]
If I understand correctly, this is the motivating example where the second IconButton gets inlined but the first one doesn't.

```js
<ButtonedListGroupItem
  key={key}
  className="clean"
  buttons={<>
    <IconButton 
      title="Save Edits"
      icon="save"
      family="r"
      onClick={this.save}
    />
    <IconButton 
      title="Cancel Editing"
      icon="ban"
      onClick={this.cancel}
    />
  </>}
/>
```

So what we want is to detect when there's a list of nearby elements that are of the same shape and if one breaks, then also break the others. This is also not an issue only for JSX but for all the constructs in prettier.

The reason it hasn't been fixed yet is that I don't know how to make it happen efficiently and in a way that would be good. It would also likely play funnily with diffs when if you change one element, then you may cause all the others to expand but that's only one of the side effect I can think about right now.

We've got to make a shitty trade-off here but I believe that the status quo is better than the solution in this PR. And I can't give you some actionable next steps to fix this problem because I don't know how to do it myself.

It's not impossible that this issue will never be solved and while that's unfortunate, I believe that prettier provide enough value that it's okay.
